### ISSUE TITLE
Paragraph in list element is not converted properly

### ISSUE THREAD

**Issue by pkruithof** [thephpleague/html-to-markdown#47]
HTML:

``` html
<li>
  <h3>Header</h3>
  <p>Description</p>
</li>
```

Expected:

``` md
- ### Header
  Description
```

Actual:

``` md
- ### Header
Description
```

Which converted back to HTML results in this:

``` html
<ul>
  <li>
    <h3>Header</h3>
  </li>
</ul>
<p>Description</p>
```

**Comment by colinodell** [comment]
Thanks for reporting this!

Because we prefer adhering to the CommonMark standard as much as possible, I think the proper output should be:

~~~markdown
- ### Header

  Description
~~~

([see dingus](http://spec.commonmark.org/dingus/?text=-%20%23%23%23%20Header%0A%0A%20%20Description%0A%0A--------------------%0A%0A-%20%23%23%23%20Header%0A%20%20Description))

**Comment by pkruithof** [comment]
Seems reasonable. I took a stab at solving this problem yesterday, but I descended quickly into a much bigger focus. You see I think the approach that the conversion method uses can be a bit flawed: with inline replacement of node contents you loose some of the original context. For instance, when converting a list, the contents of the `<ul>` node (after it has finished traversing the `<li>` nodes) look something like this:

```
"\n  - item 1 \n  - item 2"
```

When you get to nested lists, it gets more trickier:

```
"\n  - item 1 \n  - item 2 \n    - item 2.1 \n  - item 3 \n    - item 3.1"
```

I noticed that inline nodes sometimes have a leading and/or trailing space. This is because whitespace in html is resolved into a `#text` node with contents like this: `\n         `, which is replaced by a single space.

This means that this:

```html
<ul>
    <li>Item 1</li>
</ul>
```

produces different Markdown than this:

```html
<ul><li>Item 1</li></ul>
```

This got me into a whole lot of trouble when I wanted to indent an entire nested `<ul>`, as whitespace matters a lot in Markdown. You can see why indenting these list contents with 2 spaces is troublesome:

```
"\n  - item 1 \n  - item 2 \n    - item 2.1 \n  - item 3 \n    - item 3.1"
```

Now add a different use case: a `<br>` tag inside the nested `<li>`'s content:

```html
<li>
  Item 1
  <ul>
    <li>
      Item 2<br>
      Subtext
    </li>
  </ul>
</li>
```

```
* Item 1 \n  * Item 2 \nSubtext
```

And now indenting becomes even harder.

Maybe these cases above can be solved with the current approach, but hopefully you see where it can lead to. Typically, HTML that is to be converted into Markdown is not your own HTML, so you get all these weird cases of tag soup.

I'm thinking about a different approach here, more like a tree walker. Using the DOM is a smart idea, I think. But I would put the converted Markdown into new nodes (not DOM nodes, but separate classes or something), and let each type of element handle its own conversion. That way a list block element can indent its contents if it is nested, without anything inside the list knowing about this (and thus having to handle these edge cases).

I think this way it will be much easier to adhere to the CommonMark standard (also a good idea IMO), maybe even use their test suite to test the whole thing and try to obtain full support for it.

What do you think?

**Comment by colinodell** [comment]
Thanks for the detailed report!  You're right, the current implementation isn't great at tracking the current state and formatting the Markdown accordingly.

Converting it to a node tree would definitely help.  That would require significant work though, and I unfortunately don't have the time to commit to that :-/

Perhaps we can find a simpler solution.  For example, maybe we have some variable or object which tracks the current nesting level, and we can use that to calculate the indentation?

**Comment by colinodell** [comment]
Is the list issue a possible duplicate of #19?

**Comment by pkruithof** [comment]
It depends on how you solve this issue, I think.

I almost got a simple fix running at first, but every time I fixed 1 test case, it broke somewhere else. Ultimately I had to add too many little conditions to get the tests passing, which is why I started thinking about the architectural approach all together.

I'll see if I can find time to start something..no promises though, such is the nature of OSS unfortunately :wink:
